File: problems-extend-adapt/3201-3300/3263_wiring-a-doubly-list-from-a-chain/solution_values.java
import java.util.ArrayList;
import java.util.List;

class Solution {

    public DoublyListNode wireDoublyList(ListNode head) {
        // Sweep one reads: the values ride out the walk in a buffer.
        List<Integer> values = new ArrayList<>();
        for (ListNode node = head; node != null; node = node.next) {
            values.add(node.val);
        }
        // Sweep two chains: every buffered value becomes a node appended to
        // the growing tail, pointing back at the node before it.
        DoublyListNode first = null;
        DoublyListNode tail = null;
        for (int value : values) {
            DoublyListNode fresh = new DoublyListNode(value);
            if (tail != null) {
                tail.next = fresh;
                fresh.prev = tail;
            } else {
                first = fresh;
            }
            tail = fresh;
        }
        return first;
    }
}

File: problems-adapt/3201-3300/3263_wiring-a-doubly-list-from-a-chain/solution_values.js
/**
 * @param {ListNode} head
 * @return {DoublyListNode}
 */
var wireDoublyList = function (head) {
    // Sweep one reads: the values ride out the walk in a buffer.
    const values = [];
    for (let node = head; node !== null; node = node.next) {
        values.push(node.val);
    }
    // Sweep two chains: every buffered value becomes a node appended to the
    // growing tail, pointing back at the node before it.
    let first = null;
    let tail = null;
    for (const value of values) {
        const fresh = new DoublyListNode(value);
        if (tail !== null) {
            tail.next = fresh;
            fresh.prev = tail;
        } else {
            first = fresh;
        }
        tail = fresh;
    }
    return first;
};
